libraries/pod/src/pod_option.rs

```diff
@@ -0,0 +1,218 @@
+//! Generic `Option` that can be used as a `Pod`s for types that can have
+//! a `None` value.
+
+use {
+    bytemuck::{Pod, Zeroable},
+    solana_program::{program_option::COption, pubkey::Pubkey},
+};
+
+/// Trait for types that can be `None`.
+///
+/// This trait is used to indicate that a type can be `None` according to a
+/// specific value.
+pub trait Nullable: Pod {
+    /// Indicates whether the value is `None` or not.
+    fn is_none(&self) -> bool;
+}
+
+/// A "pod-enabled" type that can be used as an `Option<T>` without
+/// requiring extra space to indicate if the value is `Some` or `None`.
+///
+/// This can be used when a specific value of `T` indicates that its
+/// value is `None`.
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct PodOption<T: Nullable>(T);
+
+unsafe impl<T: Nullable> Pod for PodOption<T> {}
+
+unsafe impl<T: Nullable> Zeroable for PodOption<T> {}
+
+impl<T: Nullable> PodOption<T> {
+    #[inline]
+    pub fn new(value: T) -> Self {
+        Self(value)
+    }
+
+    #[inline]
+    pub fn to_option(&self) -> Option<&T> {
+        if self.0.is_none() {
+            None
+        } else {
+            Some(&self.0)
+        }
+    }
+
+    #[inline]
+    pub fn to_option_mut(&mut self) -> Option<&mut T> {
+        if self.0.is_none() {
```

Comment:
Yeah, it feels that we need to improve a bit the way you can check/get the value of `T`. This is a bit verbose:
```rust
assert_eq!(Into::<Option<Pubkey>>::into(none_pubkey), None);
```
I could not find a way around it without having to create a variable first (which is not bad, but also not great):
```rust
let none_pubkey_option: Option<Pubkey> = none_pubkey.into();
```

It would be nice to have a simple way to do a `match` over a `PodOption`.

Reply:
@buffalojoec After our chat, I had a closer look at the `NonZero` struct and followed a bit the pattern there. It has a `get` method to get the contained value, so I added a `fn get(self) -> Option<T>` to return the value as an `Option`.

https://doc.rust-lang.org/std/num/struct.NonZero.html

It uses `from(T) -> NonZero<T>` methods and also a `new(T) -> NonZero<T>`, similar to the one that we had before but I think the `from(...)` should be sufficient.

Reply:
Sounds good to me. I like the internal getter, I just tossed in a suggestion on the naming. I know you pulled `get` from `NonZero`, but let me know what you think.